### PR TITLE
feat(client): composites

### PR DESCRIPTION
**Pull request by millsp**
Integrates https://github.com/prisma/prisma-engines/pull/2466

### CODE REVIEW

File: packages/cli/scripts/preinstall.js [hunk 1]
```diff
@@ -63,6 +64,7 @@ function parsePackageManagerName(userAgent) {
   // - https://pnpm.js.org/en/3.6/only-allow-pnpm
   // - https://github.com/cameronhunter/npm-config-user-agent-parser
   if (userAgent) {
+    // eslint-disable-next-line no-useless-escape -- ???
```
Comment (Jolg42) [inline packages/cli/scripts/preinstall.js:67]: I thought I fixed that on main, is this branch up-to-date already?

File: packages/client/scripts/colors.js [hunk 2]
```diff
@@ -3,8 +3,8 @@
 const isObject = (val) => val !== null && typeof val === 'object' && !Array.isArray(val)
 
 // this is a modified version of https://github.com/chalk/ansi-regex (MIT License)
-/* eslint-disable-next-line no-control-regex */
 const ANSI_REGEX =
+  // eslint-disable-next-line no-control-regex
```
Comment (Jolg42) [inline packages/client/scripts/colors.js:7]: I thought I fixed that on main, is this branch up-to-date already?

Comment (Jolg42) [reply packages/client/scripts/colors.js:7]: and same comments for other places as well

File: packages/generator-helper/src/dmmf.ts [hunk 3]
```diff
@@ -75,8 +75,8 @@ export namespace DMMF {
     isUnique: boolean
     isId: boolean
     isReadOnly: boolean
-    isGenerated: boolean
-    isUpdatedAt: boolean
+    isGenerated?: boolean // does not exist on type
+    isUpdatedAt?: boolean // does not exist on type
```
Comment (Jolg42) [inline packages/generator-helper/src/dmmf.ts:79]: Do you know when it exists?

File: packages/client/src/utils/getTestClient.ts [hunk 4]
```diff
@@ -26,11 +26,9 @@ const readFile = promisify(fs.readFile)
...
  */
 export async function getTestClient(schemaDir?: string, printWarnings?: boolean): Promise<any> {
-  if (!schemaDir) {
-    const callsite = parse(new Error('').stack!)
-    schemaDir = path.dirname(callsite[1].file!)
-  }
-  const schemaPath = await getRelativeSchemaPath(schemaDir)
+  const callSite = path.dirname(require.main?.filename ?? '')
+  const absSchemaDir = path.resolve(callSite, schemaDir ?? '')
+  const schemaPath = await getRelativeSchemaPath(absSchemaDir)
```
Comment (Jolg42) [inline packages/client/src/utils/getTestClient.ts:31]: Curious to know why the change was needed though it looks good

File: packages/client/src/runtime/query.ts [hunk 5]
```diff
@@ -1050,10 +1050,13 @@ function byToSelect(by: string[]): Record<string, true> {
...
 }
 
-function getDefaultSelection(outputType: DMMF.OutputType) {
+function getDefaultSelection(dmmf: DMMFHelper, outputType: DMMF.OutputType) {
   const acc = Object.create(null)
 
   for (const f of outputType.fields) {
+    if (dmmf.typeMap[(f.outputType.type as DMMF.OutputType).name] !== undefined) {
+      acc[f.name] = true
+    }
```
Comment (Jolg42) [inline packages/client/src/runtime/query.ts:1059]: Maybe add a comment about what this does?

File: packages/client/src/generation/TSClient/Count.ts [hunk 6]
```diff
@@ -84,7 +84,7 @@ class CountDelegate implements Generatable {
     if (!mapping) {
       return ''
     }
-    const model = this.dmmf.modelMap[name]
+    const model = this.dmmf.typeModelMap[name]
```
Comment (Jolg42) [inline packages/client/src/generation/TSClient/Count.ts:87]: Maybe could be better if we also rename the const? It can be confusing now I think.
From `const model` to `const modelType` maybe?

File: packages/client/src/generation/TSClient/Model.ts [hunk 7]
```diff
@@ -329,29 +342,36 @@ export class ModelDelegate implements Generatable {
...
-    if (!mapping) {
-      return ''
-    }
-    const model = this.dmmf.modelMap[name]
 
-    const actions = Object.entries(mapping).filter(
-      ([key, value]) => key !== 'model' && key !== 'plural' && key !== 'aggregate' && key !== 'groupBy' && value,
-    )
+    const mapping = this.dmmf.mappingsMap[name] ?? { model: name, plural: `${name}s` }
+    const model = this.dmmf.typeModelMap[name]
```
Comment (Jolg42) [inline packages/client/src/generation/TSClient/Model.ts:347]: Maybe could be better if we also rename the const? It can be confusing now I think.
From `const model` to `const modelType` maybe?

File: packages/client/src/generation/TSClient/TSClient.ts [hunk 8]
```diff
@@ -144,7 +144,7 @@ ${buildNFTAnnotations(engineType, platforms, relativeOutdir)}
     const collector = new ExportCollector()
 
     const commonCode = commonCodeTS(this.options)
-    const models = Object.values(this.dmmf.modelMap).reduce((acc, model) => {
+    const models = Object.values(this.dmmf.typeModelMap).reduce((acc, model) => {
```
Comment (Jolg42) [inline packages/client/src/generation/TSClient/TSClient.ts:147]: Maybe could be better if we also rename the const? It can be confusing now I think.
From `const models` to `const modelsType` maybe?

File: packages/client/src/utils/getTestClient.ts (hunk 4 above)
Comment (millsp) [reply packages/client/src/utils/getTestClient.ts:31]: I removed the stack trace as `require.main` should be faster. But the real change is about being able to get an in-memory client and specifying a relative path (see tests) or an absolute path. Either absolute or relative works thanks to `path.resolve`.

File: packages/client/src/generation/TSClient/TSClient.ts (hunk 8 above)
Comment (Jolg42) [reply packages/client/src/generation/TSClient/TSClient.ts:147]: What makes more sense to you, that looks good. Better than keeping the old const name I think

File: packages/client/src/generation/TSClient/Model.ts [hunk 9]
```diff
@@ -54,20 +54,18 @@ export class Model implements Generatable {
   }
   protected get argsTypes(): Generatable[] {
     const { mapping, model } = this
```
Comment (Jolg42) [inline packages/client/src/generation/TSClient/Model.ts:56]:
```suggestion
    const { mapping } = this
```
`model` is unused apparently

File: packages/client/src/generation/TSClient/Model.ts [hunk 10]
```diff
@@ -330,29 +343,36 @@ export class ModelDelegate implements Generatable {
...
-    const mapping = this.dmmf.mappingsMap[name]
-    if (!mapping) {
-      return ''
-    }
-    const model = this.dmmf.modelMap[name]
 
-    const actions = Object.entries(mapping).filter(
-      ([key, value]) => key !== 'model' && key !== 'plural' && key !== 'aggregate' && key !== 'groupBy' && value,
-    )
+    const mapping = this.dmmf.mappingsMap[name] ?? { model: name, plural: `${name}s` }
```
Comment (Jolg42) [inline packages/client/src/generation/TSClient/Model.ts:347]: Pluralizing is often a bad idea, we should remove it in the future (and stop adding more)

File: packages/client/src/generation/TSClient/common.ts [hunk 11]
```diff
@@ -277,7 +277,11 @@ type Without<T, U> = { [P in Exclude<keyof T, keyof U>]?: never };
  * XOR is needed to have a real mutually exclusive union type
  * https://stackoverflow.com/questions/42123407/does-typescript-support-mutually-exclusive-types
  */
-type XOR<T, U> = (T | U) extends object ? (Without<T, U> & U) | (Without<U, T> & T) : T | U;
+type XOR<T, U> =
+  T extends object ?
+  U extends object ?
+    (Without<T, U> & U) | (Without<U, T> & T)
+  : U : T
```
Comment (Jolg42) [inline packages/client/src/generation/TSClient/common.ts:284]: The original XOR was a copy pasta from stackoverflow, what is the change about? (add a comment?)

File: packages/cli/scripts/preinstall.js (hunk 1 above)
Comment (millsp) [reply packages/cli/scripts/preinstall.js]: Merged with main

File: packages/client/scripts/colors.js (hunk 2 above)
Comment (millsp) [reply packages/client/scripts/colors.js]: Merged with main that's gone now

File: packages/generator-helper/src/dmmf.ts (hunk 3 above)
Comment (millsp) [reply packages/generator-helper/src/dmmf.ts:79]: I updated with a better comment locally

File: packages/client/src/runtime/query.ts (hunk 5 above)
Comment (millsp) [reply packages/client/src/runtime/query.ts:1060]: Done

File: packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/schema.prisma [hunk 12]
```diff
@@ -34,6 +34,33 @@ model User {
...
+  boolean            Boolean
+  embedEmbedList     EmbedEmbed[]
+  requiredEmbedEmbed EmbedEmbed
+  optionalEmbedEmbed EmbedEmbed?
+}
+
+type EmbedEmbed {
+  text    String
+  boolean Boolean
 }
```
Comment (Jolg42) [inline packages/client/src/__tests__/integration/happy/exhaustive-schema-mongo/schema.prisma:64]: Would be interesting to know how much recursion should be supported / tested.
Introspection default is now -1 = infinite

**Review by Jolg42 — APPROVED** [review]
That's quite a bunch of tests!

I think it could be interesting to share the work you did with @aqrln here later

File: packages/client/src/generation/TSClient/common.ts (hunk 11 above)
Comment (millsp) [reply packages/client/src/generation/TSClient/common.ts:284]: Yep, the original type effectively broke type distribution with `(T | U)`, which is necessary for nested `XOR`s that are unions. It would in turn break in unexpected ways, so probably an issue on the TS side (since required fields would be required but show no error when not provided) - so caused by us, after all.
